### DISCUSSION TITLE
Nonlinear MPC contorl of quadrotor, Casadi

### DISCUSSION THREAD

**Discussion by aminrbspace** [casadi/casadi#3558] [Q&A]
Hi, I controlled a quadrotor using CasADi for trajectory following(X,Y,Z). Then in another scenario, I changed parameters significantly in (mass, moment inertia), now unable to control desired trajectory. Adjusted weight matrices (Q, R) and prediction horizon without success.  Appreciate any help. Thanks!

**Comment by sandeep026** [comment]
Need more information. 
1. Does the solver fail?
2. Which solver and what does solver output look like?

**Comment by aminrbspace** [comment]
Hello, I hope you are fine.

1. No
2. ipopt

For quad rotor specifications, there are two cases: (simulation time steps:
(dt=0.02))

1)  mass= .34 kg; Ixx= 321862.60*10^-9; Iyy= 305825.42*10^-9; Izz=
576255.87*10^-9;
2) mass= 2000 kg; Ixx= 12000; Iyy= 20000; Izz= 9000;

For case 1, Casadi works well by N=28, and follows my references well.
For case 2, I changed all weight matrices(Q, R) and prediction horizon in
many different values with spending a lot of time, but My model does not
follow my references and diverges

I have attached the output of two cases for your reference.

I would appreciate it if you can help me. Thank you.

Best regards,

**Reply by sandeep026** [reply]
1. ensure no plant-model mismatch
2.  try setting the prediction horizon to entire simulation time and see if the optimal trjaectory converges.
3. since IPOPT solver works fine, it has mostly to do with tuning or discrepancies in the simulation setup.

**Comment by aminrbspace** [comment]
Thank you so much for your response.

Actually, I did all these steps, but  I could not get results, and I also
changed the weighting matrices from Identity to 1e6, but no changes in
results had been observed.

Best regards,
